rdrf/rdrf/auth/password_validation.py

```diff
@@ -128,3 +135,38 @@
 
     def get_help_text(self):
         return _("You must change the password to something other than your current password.")
+
+
+class EnhancedCommonPasswordValidator:
+
+    breached_password_detection = None
+    max_breach_threshold = None
+
+    def __init__(self, breached_password_detection, max_breach_threshold=None):
+        if breached_password_detection:
+            assert max_breach_threshold, 'Invalid max_breach_threshold, expected value greater than 0'
+        self.breached_password_detection = breached_password_detection
+        self.max_breach_threshold = max_breach_threshold
+
+    def validate(self, password, user=None):
+        validated = False
+        if self.breached_password_detection:
+            # Hash the password before we send it away, so that we are sure we are not exposing the user's password.
+            sha1_password = hashlib.sha1(password.encode("utf8")).hexdigest()
```

Comment:
Not sure what to do with this one. I could take this line out and pass the plaintext password to the pwnedpasswords library (which will then hash the password with SHA1 before it sends it off to the API). I just thought it was better we hash it before we pass it on to a library which could potentially change.

Reply:
I think we can dismiss this as a false positive. According to the dataset creator & CloudFlare, it is valid to use the hash because it's only using the first 5 characters:

https://blog.cloudflare.com/validating-leaked-passwords-with-k-anonymity/
https://www.troyhunt.com/ive-just-launched-pwned-passwords-version-2/
https://github.com/lionheart/pwnedpasswords/blob/717702ea0b4c62f713cc54344ae7cacd99c65e44/pwnedpasswords/pwnedpasswords.py#L131

If the library changed to introduce an intentional security vulnerability, it wouldn't necessarily matter how you call it, as basically as soon as it's imported, it can take over the application.

If it's changed and introduced an unintentional security vulnerability (ie. searching the full hashed password rather than the first 5 characters), that would still be a problem.

The API is simple enough that I think we could implement the logic to query it ourselves.

Arguably we could even host it ourselves fairly trivially, to mitigate the concerns even more. Might have to talk about it with the team.

Reply:
As discussed:
1. Replace pwnedpasswords library with our own code (simplified)
2. Make the API endpoint configurable via env variable so that we can switch to a privately hosted API in the future
3. Make enhanced password validation opt in rather than opt out.